python-spec/src/somacore/query/fast_csr.py

```diff
@@ -0,0 +1,310 @@
+        self,
+        obs_joinids: npt.NDArray[np.int64],
+        var_joinids: npt.NDArray[np.int64],
+        pool: futures.Executor,
+    ):
+        self.obs_joinids = obs_joinids
+        self.var_joinids = var_joinids
+        self.pool = pool
+
+        self.shape: Tuple[int, int] = (len(self.obs_joinids), len(self.var_joinids))
+        self.obs_indexer: pd.Index = pd.Index(self.obs_joinids)
+        self.var_indexer: pd.Index = pd.Index(self.var_joinids)
+        self.row_length: npt.NDArray[np.int64] = np.zeros(
+            (self.shape[0],), dtype=_select_dtype(self.shape[1])
+        )
+
+        # COO accumulated chunks, stored as list of triples (row_ind, col_ind, data)
+        self.coo_chunks: List[
+            Tuple[
+                npt.NDArray[np.integer],  # row_ind
+                npt.NDArray[np.integer],  # col_ind
+                npt.NDArray[np.number],  # data
+            ]
+        ] = []
+
+    def append(
+        self,
+        row_joinids: pa.Array,
+        col_joinids: pa.Array,
+        data: pa.Array,
+    ) -> None:
+        """
+        At accumulation time, do several things:
+        * re-index to positional indices, and if possible, cast to smaller dtype
+          to minimize memory footprint (at cost of some amount of time)
+        * accumulate column counts by row, i.e., build the basis of the indptr
+        * cache the tuple of data, row, col
+        """
+        rows_future = self.pool.submit(
+            _reindex_and_cast,
+            self.obs_indexer,
+            row_joinids.to_numpy(),
+            _select_dtype(self.shape[0]),
+        )
+        cols_future = self.pool.submit(
+            _reindex_and_cast,
+            self.var_indexer,
+            col_joinids.to_numpy(),
+            _select_dtype(self.shape[1]),
+        )
+        row_ind = rows_future.result()
+        col_ind = cols_future.result()
+        self.coo_chunks.append((row_ind, col_ind, data.to_numpy()))
+        _accum_row_length(self.row_length, row_ind)
+
+    def finalize(self) -> _CSRAccumulatorFinalResult:
+        nnz = sum(len(chunk[2]) for chunk in self.coo_chunks)
+        index_dtype = _select_dtype(nnz)
+        if nnz == 0:
+            # no way to infer matrix dtype, so use default and return empty matrix
+            empty = sparse.csr_matrix((0, 0))
+            return _CSRAccumulatorFinalResult(
+                data=empty.data,
+                indptr=empty.indptr,
+                indices=empty.indices,
+                shape=(0, 0),
+            )
+
+        # cumsum row lengths to get indptr
+        indptr = np.empty((self.shape[0] + 1,), dtype=index_dtype)
+        indptr[0:1] = 0
+        np.cumsum(self.row_length, out=indptr[1:])
+
+        # Parallel copy of data and column indices
+        indices = np.empty((nnz,), dtype=index_dtype)
+        data = np.empty((nnz,), dtype=self.coo_chunks[0][2].dtype)
+
+        # empirically determined value. Needs to be large enough for reasonable
```

Comment:
Super-nit: capitalize Empirically at start of sentence

Reply:
super nit, but super easy - fixed!